// cypress-video/commands/style.js
/**
 * Adds the animation style classes to the document, such that video elements can be animated.
 * Safe to call twice on the same page visit - will not add styles twice.
 * Not used as a command by the extension itself, but I left it a command in case people want to manually add the animation classes we provide.
 */
export function addStyles() {
    if (Cypress.env('CY_VIDEO_MODE') != true) {
        return
    }

    const document = cy.state('document')

    // Don't add the styles multiple times
    if (document.body.hasAttribute('data-cypress-video-creator-styles-added')) {
        return
    }

    document.body.setAttribute('data-cypress-video-creator-styles-added', "")

    // The animation style classes
    const animations = document.createElement('style')
    animations.innerHTML = `
    @keyframes cypress-video-creator-fadeIn {
        0% {
            opacity: 0;
        }
        100% {
            opacity: 1;
        }
    }

    @keyframes cypress-video-creator-fadeOut {
        0% {
            opacity: 1;
        }
        100% {
            opacity: 0;
        }
    }

    @keyframes cypress-video-creator-expanding {
        0% {
            transform: scale(0, 0);
        }
        100% {
            transform: scale(1, 1);
        }
    }

    @keyframes cypress-video-creator-shrinking {
        0% {
            transform: scale(1, 1);
        }
        100% {
            transform: scale(0, 0);
        }
    }
    .cypress-video-creator-fade-in {
        animation: cypress-video-creator-fadeIn ease 1s;
        animation-fill-mode: forwards;
    }

    .cypress-video-creator-fade-out {
        animation: cypress-video-creator-fadeOut ease 1s;
        animation-fill-mode: forwards;
    }

    .cypress-video-creator-expand {
        animation: cypress-video-creator-expanding linear 1s;
        animation-fill-mode: forwards;
    }

    .cypress-video-creator-shrink {
        animation: cypress-video-creator-shrinking linear 1s;
        animation-fill-mode: forwards;
    }

    `;
    document.body.appendChild(animations)
}

/**
 * Colour presets.
 * Values are HTML #RGB hex codes.
 * To be expanded.
 */
export const COLOR_DICT = {
    'yellow'    : '#FFD200',
    'light blue': '#89CFF0',
    'blue'      : '#383dc9',
    'green'     : '#73A657',
    'red'       : '#de3023',
    'lime'      : '#60ed13',
    'pink'      : '#ff94e2',
    'lavender'  : '#e694ff',
    'purple'    : '#8f4ea3',
    'orange'    : '#ff9900',
    'aquamarine': '#23fac1',
    'white'     : '#ffffff',
    'black'     : '#000000',
    'light gray': '#b0b0b0',
    'dark gray' : '#4f4f4f',
    'brown'     : '#704300'
}

/**
 * Gets the full name of the animation classes that don't possibly interfere with existing classes on websites.
 * @param {string} shorthand - The shorthand that the programmer uses for the animation.
 * @returns {string} The full name of the animation class.
 */
export function getAnimationClass(shorthand) {
    return 'cypress-video-creator-' + shorthand
}

Cypress.Commands.add('addStyles', addStyles)
